Debug CI

emhane opened the pull request:
## Issue Addressed

Debug CI

## Proposed Changes

I enabled testing eip4844 FORK in beacon chain tests, and now debugging tests, remaining:
- [ ] get `release-tests-windows` to compile (@pawanjay176 will address this one)
- [x] `release-tests-ubuntu` (2 failing):
- [x] `rpc::codec::ssz_snappy::tests::test_context_bytes_v2` 
- [x] `rpc::codec::ssz_snappy::tests::test_encode_then_decode_v2` (@ackintosh has a fix for[ these 2](https://github.com/emhane/lighthouse/pull/1), Update: I committed the fix you sent me, thanks) 
- [x] `beacon-chain-tests` (2 failing): 
- [x] `store_tests::finalizes_non_epoch_start_slot` (I think this is a logic bug. Only fails for eip4844 fork, would you please debug it @realbigsean @pawanjay176)
- [x] `store_tests::weak_subjectivity_sync`
- [x] clippy
- [x] `debug-tests-ubuntu`:
- [x] `block lookup`
- [x] `sync range tests`
- [x] `cargo-udeps`

## Additional Info

Beacon chain tests worked before enabling eip4844 fork in makefile https://github.com/sigp/lighthouse/actions/runs/3980973806

emhane committed:
Fix network block_lookups test

emhane committed:
Debug tests

emhane committed:
fixup! Debug tests

emhane committed:
fixup! Debug tests

emhane committed:
Fix release test

emhane committed:
Add eip4844 fork to tests

emhane committed:
Fix weak_subjectivity_sync test

emhane committed:
Fix range sync tests

emhane committed:
fixup! Fix block lookup debug tests

emhane committed:
Fix rebase conflicts

emhane committed:
Fix rebase conflicts

emhane committed:
Fix rebase conflicts

emhane committed:
Fixes after rebasing eip4844

emhane committed:
Fix clippy

emhane committed:
Fix release tests

emhane committed:
Change CI clippy

emhane committed:
fixup! Change CI clippy

emhane commented:
How do we feel about allowing this clippy lint `clippy::uninlined-format-args` and other lints to make the original clippy CI command behave like in rust 1.66.0? Or does someone have some regex that can nicely refactor 1360 format! strings with different variable count and names, of which random sampling tells me, none follow the clippy lint.

paulhauner commented:
> How do we feel about allowing this clippy lint clippy::uninlined-format-args

Sounds good to me!

> and other lints to make the original clippy CI command behave like in rust 1.66.0?

Generally fine with this, but if we ignore anything that might be useful in the future then it might be nice to make an issue for it so we don't forget.

ackintosh commented:
#### beacon-chain-tests >  `store_tests::finalizes_non_epoch_start_slot`

https://github.com/sigp/lighthouse/actions/runs/4019502132/jobs/6906436474
> ---- store_tests::finalizes_non_epoch_start_slot stdout ----
> thread 'store_tests::finalizes_non_epoch_start_slot' panicked at 'assertion failed: `(left == right)`
>   left: `{SignedBeaconBlockHash(0x1b5b1623565a294b2d8374801e803ba6534a8ed3c9329c1e953209472f934a0f), SignedBeaconBlockHash(0xd8fa268f1e6335decadee7e61f8bb013217dd5acb5ef770740b5dc43f017754e)}`,
>  right: `{SignedBeaconBlockHash(0x1b5b1623565a294b2d8374801e803ba6534a8ed3c9329c1e953209472f934a0f), SignedBeaconBlockHash(0xb0b3170a7afb91e0ccaaa323272aa3c97e44c13ea7f4b8b30137c0776f23bba3)}`', beacon_node/beacon_chain/tests/store_tests.rs:1720:5

The failed assertion is here:

https://github.com/sigp/lighthouse/blob/690598740049234459d68fecdf9979995a92be36/beacon_node/beacon_chain/tests/store_tests.rs#L1720-L1726

<img width="1334" alt="image" src="https://user-images.githubusercontent.com/1885716/215313077-d366c544-472d-4291-b0ba-310fa233c34f.png">

From what I have seen, I think that comes from about [here](https://github.com/sigp/lighthouse/blob/b4689e20c6508e58e8245431487e6c645d386ee7/consensus/state_processing/src/state_advance.rs#L28-L45) but I haven't figured the cause out so far. 🤔 

---

**Update**: Comparing the finalized checkpoints between the forks (merge, capalla, eip4844). Seems weird that non-zero hashes appear since Slot(32) not Slot(40) on `eip4844`.

<img width="2511" alt="image" src="https://user-images.githubusercontent.com/1885716/216848594-1551fd25-c1cb-470e-8e4b-4c6ffe43fb5a.png">

I'm not 100% confident but I think the difference above comes from updating `current_justified_checkpoint`.

In [`weigh_justification_and_finalization()`](https://github.com/sigp/lighthouse/blob/20ebf1f3c111197b92e8f065880b03623645a1a4/consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs#L8), we have  following `(a)` and `(b)` to update `current_justified_checkpoint`. In the case of merge and capella, the `(b)` never happens. On the other hand in the case of eip4844 `(b)` happens and moreover, `(a)` and `(b)` happens in the same [`weigh_justification_and_finalization()`](https://github.com/sigp/lighthouse/blob/20ebf1f3c111197b92e8f065880b03623645a1a4/consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs#L8) call.

|  Updating current_*justified*_checkpoint  | Code |
| ------------- | ------------- |
| `(a)` ... with  previous epoch  | https://github.com/sigp/lighthouse/blob/20ebf1f3c111197b92e8f065880b03623645a1a4/consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs#L24-L30  |
| `(b)` ... with current epoch  | https://github.com/sigp/lighthouse/blob/20ebf1f3c111197b92e8f065880b03623645a1a4/consensus/state_processing/src/per_epoch_processing/weigh_justification_and_finalization.rs#L31-L38  |

---

(**Update on 2023-02-07**)

Wow, the bug I was looking into has been solved since emilia has rebased this. 😹 

I couldn't identify which changes solve that,  the `(b)` never happens in the case of eip4844 too now.

michaelsproul committed:
Unpin fixed-hash (#3917)

## Proposed Changes
Remove the `[patch]` for `fixed-hash`.

We pinned it years ago in #2710 to fix `arbitrary` support. Nowadays the 0.7 version of `fixed-hash` is only used by the `web3` crate and doesn't need `arbitrary`.

~~Blocked on #3916 but could be merged in the same Bors batch.~~

emhane committed:
Fix conflicts rebasing eip4844

emhane committed:
Fix conflicts rebasing eip4844

emhane committed:
Fix conflicts rebasing eip4844

emhane committed:
Fix conflicts rebasing eip4844

emhane committed:
Fix latest clippy lints

emhane committed:
fixup! Fix conflicts rebasing eip4844

paulhauner commented:
@ackintosh, nice detective work. I came across some issues similar to this over in https://github.com/ethDreamer/lighthouse/pull/9 that might be related to this. Over in that PR, I found that a change to the genesis validator set was changing the signatures that form our "randao reveals"; the psuedorandom values that we use to assign validators to attest at particular slots (aka, the "shuffling"). Because of the way our tests run with a very low validator count, the change in shuffling could result in different finality outcomes. In this case, it could be that the changing of the fork versions is changing the signatures, which is then changing the randao reveals.

It would be interesting to see if cherry-picking https://github.com/ethDreamer/lighthouse/pull/9 to this branch would solve the test failures.

ackintosh commented:
I have tested cherry-picking https://github.com/ethDreamer/lighthouse/pull/9 to this branch locally and confirmed the test has passed! The change below solves the test failures I was looking into. Thank you @paulhauner !

```diff
#[tokio::test]
async fn finalizes_non_epoch_start_slot() {
-    const HONEST_VALIDATOR_COUNT: usize = 16 + 0;
-    const ADVERSARIAL_VALIDATOR_COUNT: usize = 8 - 0;
+    const HONEST_VALIDATOR_COUNT: usize = 32 + 0;
+    const ADVERSARIAL_VALIDATOR_COUNT: usize = 16 - 0;
```

michaelsproul committed:
Update dependencies (#3946)

Resolves the cargo-audit failure caused by https://rustsec.org/advisories/RUSTSEC-2023-0010.

I also removed the ignore for `RUSTSEC-2020-0159` as we are no longer using a vulnerable version of `chrono`. We still need the other ignore for `time 0.1` because we depend on it via `sloggers -> chrono -> time 0.1`.

emhane committed:
Fix conflicts rebasing eip4844

emhane committed:
fixup! Fix latest clippy lints

emhane committed:
Remove unused dependencies

emhane committed:
fixup! Update dependencies (#3946)

emhane committed:
fixup! Add tests for blob pruning flags

emhane committed:
Fix lint for prune blobs pr

emhane committed:
Run cargo update

emhane committed:
fixup! Add tests for blob pruning flags

realbigsean commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -1070,32 +1066,27 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         match self.store.get_blobs(block_root)? {
             Some(blobs) => Ok(Some(blobs)),
             None => {
-                // Check for the corresponding block to understand whether we *should* have blobs.
```
Think this might be a rebase issue?

realbigsean replied:
updated here: https://github.com/emhane/lighthouse/pull/2/commits/28702c9d5d448ede4b2d9a2673add5921541812e

emhane replied:
Oh man, it was updated. Scary rebase issues. Thanks for keeping an eye out for them with me!

emhane replied:
Wait, at glance it looks identical to the method I coded in the check dab rpc pr, that's why I recognised it. Anyhow will merge from now on to avoid these automatic rebase problems.

realbigsean commented on lighthouse/tests/beacon_node.rs:
```diff
@@ -1357,27 +1357,27 @@ fn prune_blobs_on_startup_false() {
+        .flag("epochs-per-blob-prune", Some("5"))
         .run_with_zero_port()
-        .with_config(|config| assert!(!config.epochs_per_blob_prune == 5));
+        .with_config(|config| assert!(config.store.epochs_per_blob_prune == 5));
```
oh dang were these tests wrong before? nice catch!

realbigsean commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -131,19 +132,23 @@ pub enum BlockProposalContents<T: EthSpec, Payload: AbstractExecPayload<T>> {
     },
 }
 
+pub struct BlockProposalBlobsContents<T: EthSpec> {
+    pub kzg_commitments: KzgCommitments<T>,
+    pub blobs: Blobs<T>,
+}
+
+pub struct BlockProposalContentsDeconstructed<T: EthSpec, Payload: AbstractExecPayload<T>> {
+    pub payload: Payload,
+    pub blobs_content: Option<BlockProposalBlobsContents<T>>,
+}
```
I'm not seeing where these two are used?

realbigsean replied:
updated: https://github.com/emhane/lighthouse/pull/2/commits/8f9c5cfca9e4478530a92207a5fdfc4c06bf6564

realbigsean commented on beacon_node/network/src/sync/range_sync/chain.rs:
```diff
@@ -615,6 +615,7 @@ impl<T: BeaconChainTypes> SyncingChain<T> {
     ///
     /// If a previous batch has been validated and it had been re-processed, penalize the original
     /// peer.
+    #[allow(clippy::modulo_one)]
```
nice, agree with this in case we increase batch size in the future

realbigsean commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -534,21 +560,23 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
-                && super::range_sync::EPOCHS_PER_BATCH == 1,
-            "To deal with alignment with 4844 boundaries, batches need to be of just one epoch"
-        );
+    pub fn batch_type(&self, _epoch: types::Epoch) -> ByRangeRequestType {
```
The `_` prefix ends up making this param look like it's never used, so maybe we can `#[allow(unused)]` instead

realbigsean replied:
Updated here: https://github.com/emhane/lighthouse/pull/2/commits/fc2d07b4e393c8e0c9331d83e8381262531692b1

realbigsean commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -534,21 +560,23 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
+        if super::backfill_sync::BACKFILL_EPOCHS_PER_BATCH * super::range_sync::EPOCHS_PER_BATCH
+            != 1
+        {
+            panic!(
```
I think this was written using a `const` before to get a compile time panic, whereas this would give us a run-time panic which we don't want. I'm guessing this was to fix a lint, maybe there's another way to work around this.

realbigsean replied:
updated here: https://github.com/emhane/lighthouse/pull/2/commits/4c3561dcaf8ddeb7cef86f329a386615adfad9eb

realbigsean commented on beacon_node/lighthouse_network/src/rpc/protocol.rs:
```diff
@@ -409,14 +409,14 @@ impl ProtocolId {
-                Protocol::BlobsByRange | Protocol::BlobsByRoot => return true,
-                _ => return false,
-            },
+            Version::V2 => matches!(
```
this looks like an improvement to me, I think we should maybe go further though and make this a complete match. This would force us to consciously decide whether any added protocols should have context bytes.

realbigsean replied:
updated here: https://github.com/emhane/lighthouse/pull/2/commits/ad9af6d8b128215c4a86f58f1ab5a72ac60fb3e7

realbigsean committed:
merge upstream, add back `get_blobs` logic

realbigsean committed:
allow unused

realbigsean committed:
complete match for `has_context_bytes`

realbigsean committed:
remove unused structs

realbigsean committed:
make batch size check compile time panic

realbigsean committed:
make batch size check compile time panic

realbigsean committed:
Revert "make batch size check compile time panic"

This reverts commit 68f2484efcccd3a51fd1f92356cf95671d0c9eb1.

realbigsean committed:
fix a couple new lints

realbigsean commented:
Awesome job guys! 

I addressed a bunch of my own comments [here](https://github.com/emhane/lighthouse/pull/2), and resolved the merge conflict this has. Feel free to merge it into this branch if you agree with the changes

A big portion of the diff here is from the refactor to allow the beacon chain harness to be started up with a `SystemTimeSlotClock`. From what I can tell this became necessary due to the inclusion of an `Arc<BeaconChain>` in the `SyncNetworkContext`.  I actually am not sure including that reference is a good idea generally. I was wondering if we should instead crate something akin to the `EngineState` to track the data availability boundary in the network layer. This should probably be done in a separate PR though. 

The sync tests where we are now using the `SystemTimeSlotClock` might actually  benefit from the `TestingSlotClock` when we add 4844 related tests because we could more easily manipulate the current slot when we want to sync across the data availability boundary.

~~With that said, there might be use cases for the `SystemTimeSlotClock` harness elsewhere or in the future so I am sort of thinking we might as well keep it here.~~

I asked Paul/Michael about the potential usefulness of a beacon chain harness with the `SystemTimeSlotClock` and it seems like needing the `SystemTimeSlockClock` with the harness is usually a symptom the harness shouldn't be used generally. It tends to lead to very unstable tests. So with that, I think we should actually revert that change (sorry 😭 ) and I'm ok with merging this with those sync tests failing or commented out so that we can resolve them in a separate PR that handles getting `data_availability_boundary` information correctly. I can work on reverting it in my PR tomorrow if you aren't able to get to it today.

emhane committed:
Merge pull request #2 from realbigsean/sean-debug-ci

Debug CI Sean's PR feedback

emhane committed:
Fix merge conflicts with eip4844

emhane committed:
Remove unused error

emhane commented:
> Awesome job guys!
> 
> I addressed a bunch of my own comments [here](https://github.com/emhane/lighthouse/pull/2), and resolved the merge conflict this has. Feel free to merge it into this branch if you agree with the changes
> 
> A big portion of the diff here is from the refactor to allow the beacon chain harness to be started up with a `SystemTimeSlotClock`. From what I can tell this became necessary due to the inclusion of an `Arc<BeaconChain>` in the `SyncNetworkContext`. I actually am not sure including that reference is a good idea generally. I was wondering if we should instead crate something akin to the `EngineState` to track the data availability boundary in the network layer. This should probably be done in a separate PR though.
> 
> The sync tests where we are now using the `SystemTimeSlotClock` might actually benefit from the `TestingSlotClock` when we add 4844 related tests because we could more easily manipulate the current slot when we want to sync across the data availability boundary.
> 
> ~With that said, there might be use cases for the `SystemTimeSlotClock` harness elsewhere or in the future so I am sort of thinking we might as well keep it here.~
> 
> I asked Paul/Michael about the potential usefulness of a beacon chain harness with the `SystemTimeSlotClock` and it seems like needing the `SystemTimeSlockClock` with the harness is usually a symptom the harness shouldn't be used generally. It tends to lead to very unstable tests. So with that, I think we should actually revert that change (sorry 😭 ) and I'm ok with merging this with those sync tests failing or commented out so that we can resolve them in a separate PR that handles getting `data_availability_boundary` information correctly. I can work on reverting it in my PR tomorrow if you aren't able to get to it today.

No need to cry (lol), I jumped on this task, that's the bazaar style development :) Thanks for checking with Michael and Paul.

emhane committed:
Disable use of system time in tests

emhane committed:
Comment out tests that use system time

emhane committed:
fixup! Disable use of system time in tests

realbigsean committed:
add v3 to capabilities

michaelsproul committed:
Upgrade sqlite3

emhane committed:
Fix lint

emhane committed:
Better fix for debug tests

realbigsean commented on beacon_node/beacon_chain/src/builder.rs:
```diff
@@ -954,13 +954,14 @@ where
     TColdStore: ItemStore<TEthSpec> + 'static,
     TEth1Backend: Eth1ChainBackend<TEthSpec> + 'static,
     TEthSpec: EthSpec + 'static,
+    TSlotClock: SlotClock + 'static,
```
I think we should git rid of this generic because this impl block just adds a `testing_slot_clock` method so doesn't really make sense for it to be usable with the system time slot clock

realbigsean reviewed (approved):
One smol change but looks good otherwise!

emhane committed:
Remove unused generic

realbigsean commented:
🎉
